src/error.rs

```diff
@@ -693,6 +698,49 @@ impl WriteFailure {
     }
 }
 
+/// An error that occurred during a GridFS operation.
+#[derive(Clone, Debug)]
+#[allow(missing_docs)]
+#[non_exhaustive]
+pub enum GridFsErrorKind {
+    /// The file with the given identifier was not found.
+    #[non_exhaustive]
+    FileNotFound { identifier: GridFsFileIdentifier },
+
+    /// The file with the given revision was not found.
+    #[non_exhaustive]
+    RevisionNotFound { revision: i32 },
+
+    /// The chunk at index 'n' was missing.
+    #[non_exhaustive]
+    MissingChunk { n: u32 },
+
+    /// The chunk was the incorrect size.
+    #[non_exhaustive]
+    WrongSizeChunk {
+        actual_size: u32,
+        expected_size: u32,
+    },
+
+    /// An incorrect number of chunks was present for the file.
+    #[non_exhaustive]
+    WrongNumberOfChunks {
+        actual_number: u32,
+        expected_number: u32,
+    },
+}
+
+/// An identifier for a file stored in a GridFS bucket.
+#[derive(Clone, Debug)]
+#[non_exhaustive]
+pub enum GridFsFileIdentifier {
```

Comment:
note: I thought about whether it would make sense to accept this type in the API methods (I guess`String` and `Bson` could implement `Into<GridFsFileIdentifier>` so users wouldn't have to construct identifiers?) and avoid having separate methods for download by filename and ID. 
but I think the separate named methods is probably a little nicer in terms of discovering the allowed functionality, and it allows us to separately document the behavior of querying by filename vs ID. so I think the current API is fine.